### ISSUE TITLE
k6/html: Have a way to get the selection out of an element

### ISSUE THREAD

**Issue by mstoykov** [grafana/k6#2683]
### Feature Description

As reported in https://community.k6.io/t/create-selection-from-element/4643 it is comment in jquery to do [`selection.each`](https://k6.io/docs/javascript-api/k6-html/selection/selection-each/) and then get the selection of the element that `each` will give you and continue selecting.

This is currently not possible in k6, but we do have the selection of the element as part of the element so it shouldn't be that hard.

This likely though needs some though on the API

### Suggested Solution (optional)

We can just have new [Element](https://k6.io/docs/javascript-api/k6-html/element/) method `selection()` and be done with it. But it likely won't work a lot like jquery, how important is that is different question.

We can defined `$` to mean something in k6 - maybe by calling a method on `k6/html`. And then use that :shrug:.

But it will be confusing as `$('li')` will not really mean anything as we don't know which page to query.

Maybe it can only be difined in `each` :shrug:

**Comment by Azhovan** [comment]
@mstoykov i would like to work on this issue. can you please assign it to me?

**Comment by allansson** [comment]
> The $ symbol would be a property of the Element type, not a global variable or a function, and it would take a selector string as an argument and return a Selection object that contains the element itself and any other elements that match the selector. e.g. if element is an Element that represents a `div` element, then element.$('p') would return a Selection object that contains all the `p` elements inside the `div` element.

I've never seen a `$` being used in a method/property name unless to mark it as an observable or something internal/hidden. And, in my opinion, it reads horribly. It's hard to for the eye to parse and the non-descriptive name means I would have to look it up in the docs in order to understand what it does.

> We can defined $ to mean something in k6 - maybe by calling a method on k6/html. And then use that 🤷.
>
> But it will be confusing as $('li') will not really mean anything as we don't know which page to query.

I see no need to try and be jQuery. We have the Selection API. Reading the forum post, I don't think the author even expects that the API would be using `$`. They're just using jQuery as a reference.

Why not just a method `Selection.fromElement(element: Element)`?:

```ts
doc.find("li")
  .each((_, element) => {
    const listEl = Selection.fromElement(element)
      .closest('ul')
  })
```

We're not married to WHATWG or W3C either, so maybe `Element extends Selection`?

```ts
doc.find("li")
  .each((_, element) => {
    const listEl = element.closest("ul")
  })
```

 @mstoykov's alternative is also good:

```ts
element.selection()
element.toSelection()
element.asSelection()
```

**Comment by w1kman** [comment]
Would be neat if DOM traversal in `k6/html´ and `k6/experimental/browser` had similar APIs.

**Comment by mstoykov** [comment]
@w1kman can you expand on this, please.

**Comment by Azhovan** [comment]
@w1kman 

```javascript 
import http from 'k6/http';
import { Browser } from 'k6/experimental/browser';

// Using k6/html module
export default function () {
  let res = http.get('https://abcd.com');
  // ...
  // ...  
  let dom = res.html();
  let title = dom.querySelector('title');
  let description = dom.querySelector('meta[name="description"]');
  console.log('Title: ' + title.text());
  console.log('Description: ' + description.attr('content'));
}

// Using k6/experimental/browser module
export default function () {
  let browser = new Browser();
  let page = browser.open('https://abcd.com');
  page.wait();
  let title = page.querySelector('title');
  let description = page.querySelector('meta[name="description"]');
  console.log('Title: ' + title.text());
  console.log('Description: ' + description.attr('content'));
  browser.close();
}

```
I think you mean the code are very similar the only diff is between ` let dom = res.html();` and `let page = browser.open(...)`

**Comment by Azhovan** [comment]
what do you think @mstoykov ?

**Comment by w1kman** [comment]
@Azhovan @mstoykov 

In `k6-browser` the user can use `page.locator(<css-selector/xpath>)` to find elements in the current `Document`. It would be great if the custom selectors in `k6-browser` and `k6/http. parseHTML` would return the same-ish (`k6-browser` can interact with the DOM, so I'd expect it to have an extended `Element` API) `Element`.

Personally I'd rather use vanilla `Document` and `Element`, but for those who want synthetic sugar, it would be nice if they didn't have to learn two different flavors to use `k6`.

**Comment by Azhovan** [comment]
@w1kman @mstoykov 
I just created a PR!

**Comment by Azhovan** [comment]
@mstoykov 
i think we can close this?

**Comment by mstoykov** [comment]
@Azhovan, ah yeah, I thought it was marked in the PR/commit, but apparently not.

closed with #3519
